src/julia_internal.h

```diff
@@ -289,7 +289,7 @@ JL_DLLEXPORT jl_value_t *jl_gc_alloc(jl_ptls_t ptls, size_t sz, void *ty);
 #  define jl_gc_alloc(ptls, sz, ty) jl_gc_alloc_(ptls, sz, ty)
 #endif
 
-#define jl_buff_tag ((uintptr_t)0x4eade800)
+#define jl_buff_tag ((uintptr_t)0x4eade000)
```

Comment:
I assume this is not needed anymore?

Reply:
This is something that I introduced for the reasons mentioned in an earlier comment. Basically, it is possible for a spurious reference to an object with the type `jl_buff_tag` to occur with conservative scanning; such an object must not be passed to `jl_gc_mark_queue_obj()` and similar functions that try to treat the header word as a `jl_datatype_t *` that they can derefence. So we want to make sure that `jl_buff_tag` cannot accidentally be mistaken for a value that can legitimately occur in the object header (which is extremely unlikely, but at least theoretically possible). By making it a multiple of `GC_PAGE_SZ`, we avoid that problem.

Reply:
Yes, but you should be relying on that to distinguish allocated and unallocated cell anymore?
Also, just realized that I forgot to mention, even with the conservative scan, any code that is julia-aware has to use the GC frame explicitly. More specifically, any code that creates and access julia array in C code has to do that since otherwise the array object itself might not be anywhere on the stack while the buffer is still being used which will cause 1) (from the current implementation) the buffer won't be scanned and 2) the buffer itself doesn't even contain enough information about what's in there.

Reply:
> Yes, but you should (edit: NOT) be relying on that to distinguish allocated and unallocated cell anymore?

Hmm. I was under the impression that objects with `jl_buff_tag` can still happen to occur as a live object. While they get special treatment in the mark loop, I believed that they can still occur in other contexts (e.g. when allocated off `pool->newpages`) in ways that make them indistinguishable from normal allocations.

> Also, just realized that I forgot to mention, even with the conservative scan, any code that is julia-aware has to use the GC frame explicitly. More specifically, any code that creates and access julia array in C code has to do that since otherwise the array object itself might not be anywhere on the stack while the buffer is still being used which will cause 1) (from the current implementation) the buffer won't be scanned and 2) the buffer itself doesn't even contain enough information about what's in there.

I think this is something that simply needs to be documented. Namely, that arrays need to be kept visible to the Julia GC even if they are not visible on the stack anymore if you are accessing its elements. E.g. using `JL_GC_PUSH1()` and friends, storing them in a Julia object  or a root while accessing elements, or some such. But yes, I think this is something that requires explicit documentation.

Reply:
> Hmm. I was under the impression that objects with jl_buff_tag can still happen to occur as a live object.

Ah, so what you want to distinguish `jl_buff_tag` from a real tag reliably. I though you needed to use this to distinguish between a unallocated and allocated cell. This use is fine then, but it should be documented near the definition of the `jl_buff_tag`, and possibly a "see also" here as well.

Reply:
Sounds good. :)

Reply:
> and possibly a "see also" here as well.

Oh, and by "here" I mean where you check that in the code. I thought that's where this comment thread is....